core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -35,63 +32,65 @@
-
-    // Set up zookeeper specific functionality.
+  protected FaultTolerantAlluxioMasterProcess(JournalSystem journalSystem,
+      PrimarySelector leaderSelector) {
+    super(journalSystem);
     try {
-      // InetSocketAddress.toString causes test issues, so build the string by hand
-      String zkName = NetworkAddressUtils.getConnectHost(ServiceType.MASTER_RPC) + ":"
-          + getRpcAddress().getPort();
-      String zkAddress = Configuration.get(PropertyKey.ZOOKEEPER_ADDRESS);
-      String zkElectionPath = Configuration.get(PropertyKey.ZOOKEEPER_ELECTION_PATH);
-      String zkLeaderPath = Configuration.get(PropertyKey.ZOOKEEPER_LEADER_PATH);
-      mLeaderSelectorClient =
-          new LeaderSelectorClient(zkAddress, zkElectionPath, zkLeaderPath, zkName);
-
-      // Check that the journal has been formatted.
-      MasterUtils.checkJournalFormatted();
+      stopServing();
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
+    mLeaderSelector = Preconditions.checkNotNull(leaderSelector, "leaderSelector");
+    mServingThread = null;
   }
 
   @Override
   public void start() throws Exception {
+    mJournalSystem.start();
     try {
-      mLeaderSelectorClient.start();
+      mLeaderSelector.start(getRpcAddress());
     } catch (IOException e) {
       LOG.error(e.getMessage(), e);
       throw new RuntimeException(e);
     }
 
-    Thread currentThread = Thread.currentThread();
-    mLeaderSelectorClient.setCurrentMasterThread(currentThread);
-    boolean started = false;
-
     while (!Thread.interrupted()) {
-      if (mLeaderSelectorClient.isLeader()) {
-        stopServing();
-        stopMasters();
-
-        startMasters(true);
-        started = true;
-        startServing("(gained leadership)", "(lost leadership)");
+      if (mLeaderSelector.isPrimary()) {
+        if (mServingThread == null) {
+          LOG.info("Transitioning from secondary to primary");
+          mJournalSystem.setMode(Mode.PRIMARY);
+          stopMasters();
+          LOG.info("Secondary stopped");
+          startMasters(true);
+          mServingThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+              startServing("(gained leadership)", "(lost leadership)");
+            }
+          }, "MasterServingThread");
+          mServingThread.start();
+          LOG.info("Primary started");
+        }
       } else {
         // This master should be standby, and not the leader
-        if (isServing() || !started) {
+        if (mServingThread != null) {
+          LOG.info("Transitioning from primary to secondary");
           // Need to transition this master to standby mode.
+          mServingThread.interrupt();
+          mServingThread.join();
+          mServingThread = null;
           stopServing();
           stopMasters();
-
+          mJournalSystem.setMode(Mode.SECONDARY);
+          LOG.info("Primary stopped");
           startMasters(false);
```

Comment:
The `Mode` enum is currently entrenched in the `JournalSystem` API. We could introduce a separate mode enum for masters, but I'd rather not make that change as a part of this PR